### PR TITLE
bugfix/set_discrete_sensor_reading

### PR DESCRIPTION
**Pull request by JuneZhou19**
@turtle-fly @fub2 @xiar @MarkMa001 @PaynePei

### CODE REVIEW

**Comment by xiar** [comment]
Can one of the admins verify this patch?

**Comment by JuneZhou19** [comment]
It's passed tests with all available sensor/sel/help/history commands in dell_c6320.

**Comment by emcinfrasim** [comment]
```
******* IDIC_pull_request_trigger ******* 
TestName: T43943_idic_IPMISIMHelpAccessible 
ErrorMessage: BMC IPMI_SIM of a node not accessible. Rack is vRack1, Node is vbmc_quanta_d51_pr.ova_1470667071.13, BMC IP is: 192.168.133.85,  BMC IPMI_SIM of a node not accessible. Rack is vRack1, Node is vbmc_quanta_d51_pr.ova_1470667087.96, BMC IP is: 192.168.134.180 
TestName: T46139_idic_IPMISIMSensorAccessible 
ErrorMessage: BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_d51_pr.ova_1470667071.13, BMC IP is: 192.168.133.85,  BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_d51_pr.ova_1470667087.96, BMC IP is: 192.168.134.180 
TestName: T46139_idic_IPMISIMSensorAccessible 
ErrorMessage: BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_d51_pr.ova_1470667071.13, BMC IP is: 192.168.133.85,  BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_d51_pr.ova_1470667087.96, BMC IP is: 192.168.134.180 
```

**Comment by JuneZhou19** [comment]
retest this please

**Comment by emcinfrasim** [comment]
```
******* IDIC_pull_request_trigger ******* 
TestName: T43943_idic_IPMISIMHelpAccessible 
ErrorMessage: BMC IPMI_SIM of a node not accessible. Rack is vRack1, Node is vbmc_quanta_t41_pr.ova_1470707440.61, BMC IP is: 192.168.136.7,  BMC IPMI_SIM of a node not accessible. Rack is vRack1, Node is vbmc_quanta_t41_pr.ova_1470707423.74, BMC IP is: 192.168.136.135 
TestName: T46139_idic_IPMISIMSensorAccessible 
ErrorMessage: BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_t41_pr.ova_1470707440.61, BMC IP is: 192.168.136.7,  BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_t41_pr.ova_1470707423.74, BMC IP is: 192.168.136.135 
TestName: T46139_idic_IPMISIMSensorAccessible 
ErrorMessage: BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_t41_pr.ova_1470707440.61, BMC IP is: 192.168.136.7,  BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_quanta_t41_pr.ova_1470707423.74, BMC IP is: 192.168.136.135 
```

**Comment by JuneZhou19** [comment]
retest this please

**Comment by emcinfrasim** [comment]
```
******* IDIC_pull_request_trigger ******* 
TestName: T43943_idic_IPMISIMHelpAccessible 
ErrorMessage: BMC IPMI_SIM of a node not accessible. Rack is vRack1, Node is vbmc_dell_c6320_pr.ova_1470713118.16, BMC IP is: 192.168.136.161,  BMC IPMI_SIM of a node not accessible. Rack is vRack1, Node is vbmc_dell_c6320_pr.ova_1470713135.38, BMC IP is: 192.168.136.179 
TestName: T46139_idic_IPMISIMSensorAccessible 
ErrorMessage: BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_dell_c6320_pr.ova_1470713118.16, BMC IP is: 192.168.136.161,  BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_dell_c6320_pr.ova_1470713135.38, BMC IP is: 192.168.136.179 
TestName: T46139_idic_IPMISIMSensorAccessible 
ErrorMessage: BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_dell_c6320_pr.ova_1470713118.16, BMC IP is: 192.168.136.161,  BMC IPMI_SIM sensor command is not accessible. Rack is vRack1, Node is vbmc_dell_c6320_pr.ova_1470713135.38, BMC IP is: 192.168.136.179 
```

File: vcompute/common/scripts/ipmi_sim/sdr.py [hunk 1]
```diff
@@ -29,23 +29,40 @@ def build_sensors(name, ID, mc, value, tp):
...
+    :return:
+    """
+
     result = common.send_ipmitool_command('raw', '0x04', '0x2d',
                                           hex(sensor_num))
+    if result == -1:
+        return 0
     try:
-        value = result.split()[0]
+	value = result.split()[0]
```
Comment (turtle-fly) [inline vcompute/common/scripts/ipmi_sim/sdr.py]: 3 spaces

File: vcompute/common/scripts/ipmi_sim/sdr.py [hunk 2]
```diff
@@ -29,23 +29,40 @@ def build_sensors(name, ID, mc, value, tp):
...
     except AttributeError:
-        return 0  # raw data can't be got, set raw value as 0
-    info = "sensor num: {0} value: 0x{1}".format(hex(sensor_num), value)
-    common.logger.info(info)
-    return int(value, 16)
+	    return 0 # raw data can't be got, set raw value as '0'
```
Comment (turtle-fly) [inline vcompute/common/scripts/ipmi_sim/sdr.py]: 3 spaces from last indent

File: vcompute/common/scripts/ipmi_sim/sdr.py [hunk 3]
```diff
@@ -97,6 +113,13 @@ def parse_sdrs():
         fd.seek(offset+13)
         event_type = ord(fd.read(1))
 
+	if event_type == 0x0:
```
Comment (turtle-fly) [inline vcompute/common/scripts/ipmi_sim/sdr.py]: one space mismatch as above

File: vcompute/common/scripts/ipmi_sim/sdr.py [hunk 4]
```diff
@@ -249,9 +272,12 @@ def parse_sdrs():
 
         # set sensor units 1 byte
         sensor_obj.set_su1(sensor_su1)
-        if sensor_su1 >> 6 != 0:
-            raw_value = struct.unpack('b', chr(sensor_value))[0]
-            sensor_obj.set_raw_value(raw_value)
+
+	#comment this raw value out since we have no clue
+	#how this sensor unit bit [7:6] impacts sensor reading.
```
Comment (turtle-fly) [inline vcompute/common/scripts/ipmi_sim/sdr.py]: indent~

File: vcompute/common/scripts/ipmi_sim/sensor.py [hunk 5]
```diff
@@ -47,13 +48,35 @@
...
+                                       self.expect_type))
+        return wrapper
 
 class Sensor:
     def __init__(self, name, ID, value, tp):
         self.name = name
         self.ID = ID
         self.tp = tp
         self.lock = threading.Lock()
+	self.lock_sensor_write = threading.Lock()
```
Comment (turtle-fly) [inline vcompute/common/scripts/ipmi_sim/sensor.py]: indent~

File: vcompute/common/scripts/ipmi_sim/sensor.py [hunk 6]
```diff
@@ -327,15 +421,15 @@ def output_info(self):
         # Sensor ID
         info += "| {0:<10}".format(hex(self.ID))
         # sensor value
-        value = hex(self.value)
         if self.get_event_type() == 'threshold':
             value = "%.3f" % self.get_reading_factor()[0](self.value)
-        info += "| {0:<10}".format(value)
+	elif self.get_event_type() == 'discrete':
```
Comment (turtle-fly) [inline vcompute/common/scripts/ipmi_sim/sensor.py]: indent~

File: vcompute/common/scripts/ipmi_sim/sensor.py [hunk 7]
```diff
@@ -327,15 +421,15 @@ def output_info(self):
...
+            value = self.value
 
+	info += "| {0:<10}".format(value)
```
Comment (turtle-fly) [inline vcompute/common/scripts/ipmi_sim/sensor.py]: indent~

**Comment by turtle-fly** [comment]
@JuneZhou19 OK with the overall design.
Please fix the indent issues for my obsessive-compulsive disorder :-)

**Comment by JuneZhou19** [comment]
@turtle-fly good catch Forrest. Fixed indents.

**Comment by fub2** [comment]
+1！

**Comment by sharkconi** [comment]
+1

**Pull request merged**